### PR TITLE
Add context flag to server command

### LINKED ISSUE #39: Add support to change context name on kubeconfig
I would love to be able to define the name of the context created in the `kubeconfig` file.

I have something already working, but in the PR template says that I should open an issue first to comment and discuss the functionality.

## Current Behaviour
Today the generated `kubeconfig` file looks like this

```
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: <redacted>
    server: https://<redacted>:6443
  name: default
contexts:
- context:
    cluster: default
    user: default
  name: default
current-context: default
kind: Config
preferences: {}
users:
- name: default
  user:
    password: <redacted>
    username: admin
```

## Possible Solution
I would love to be able to define something different to `default` on the values for all the values related to `cluster`, `user` and `context`

Something like this:

```
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: <redacted>
    server: https://<redacted>:6443
  name: homelab
contexts:
- context:
    cluster: homelab
    user: homelab
  name: homelab
current-context: homelab
kind: Config
preferences: {}
users:
- name: homelab
  user:
    password: <redacted>
    username: admin
```

## Context
- Be able to install several k3s clusters with k3sup from the same computer and merge all to the `$HOME/.kube/config` file without conflicts
- Have a more meaningful name when changing contexts

### PR DESCRIPTION
**Pull request by blackjid**
## Description
This flag can be used to define what name to use as context in the
generated kubeconfig file. The default is default.

Closes #39 

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

## How Has This Been Tested?
I've tested it running the command with the `--context` flag and a combinations of `--local-path` and `--merge` flags

This is the default behaviour
```
./k3sup install --ip $SERVER_IP --user ubuntu --ssh-key ~/.ssh/id_ed25519 --skip-install

KUBECONFIG="./kubeconfig" kubectl config view
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: DATA+OMITTED
    server: https://192.168.86.59:6443
  name: default
contexts:
- context:
    cluster: default
    user: default
  name: default
current-context: default
kind: Config
preferences: {}
users:
- name: default
  user:
    password: 5ceb3a3e93621d265fd147929f3ace84
    username: admin
```

Passing the `--context` flag
```
./k3sup install --ip $SERVER_IP --user ubuntu --ssh-key ~/.ssh/id_ed25519 --skip-install --context homelab

KUBECONFIG="./kubeconfig" kubectl config view
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: DATA+OMITTED
    server: https://192.168.86.59:6443
  name: homelab
contexts:
- context:
    cluster: homelab
    user: homelab
  name: homelab
current-context: homelab
kind: Config
preferences: {}
users:
- name: homelab
  user:
    password: 5ceb3a3e93621d265fd147929f3ace84
    username: admin
```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by alexellis** [comment]
Please show evidence of the testing and the file between runs with default and override

**Comment by blackjid** [comment]
I have updated the PR description.

**Comment by alexellis** [comment]
What about if I set the flag to `""`?

**Comment by blackjid** [comment]
Well in that case `default` is replaced by `""` do you want to handle that case?

```
./k3sup install --ip $SERVER_IP --user ubuntu --ssh-key ~/.ssh/id_ed25519 --context "" --skip-install

KUBECONFIG="./kubeconfig" kubectl config view
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: DATA+OMITTED
    server: https://192.168.86.59:6443
  name: ""
contexts:
- context:
    cluster: ""
    user: ""
  name: ""
current-context: ""
kind: Config
preferences: {}
users:
- name: ""
  user:
    password: 5ceb3a3e93621d265fd147929f3ace84
    username: admin
```

**Comment by blackjid** [comment]
Well that's the first try.. let me know what do you think.

File: pkg/cmd/install_test.go [hunk 1]
```diff
@@ -40,8 +63,38 @@ func Test_loadPublickeyEncrypted(t *testing.T) {
...
+	group := re.FindSubmatch(kubeconfig)
+
+	if len(group) == 0 || string(group[1]) != ip {
+		t.Errorf("Unexpected error, got: %q, want: %q.", string(group[1]), ip)
+	}
+
+	re = regexp.MustCompile(`context-test`)
+	match := re.FindAllIndex(kubeconfig, -1)
+
+	if len(match) != 6 {
```
Comment (alexellis) [inline pkg/cmd/install_test.go]: I do not like magic numbers like "6" randomly in the code, could you instead break out what you want to match and then check against the length of that?

```
want :="whatever"

if len(match) != len(want) {

}
```

File: pkg/cmd/install.go [hunk 2]
```diff
@@ -261,3 +263,16 @@ func loadPublickey(path string) (ssh.AuthMethod, func() error, error) {
 
 	return ssh.PublicKeys(signer), noopCloseFunc, nil
 }
+
+func rewriteKubeconfig(kubeconfig string, ip string, context string) []byte {
+	if context == "" {
+		context = "default"
+	}
+
+	kubeconfigReplacer := strings.NewReplacer(
```
Comment (alexellis) [inline pkg/cmd/install.go]: Nice.. I learned about this today. Thank you

File: pkg/cmd/install_test.go [hunk 3]
```diff
@@ -40,8 +63,38 @@ func Test_loadPublickeyEncrypted(t *testing.T) {
...
+		t.Errorf("Unexpected error, got: %q, want: %q.", len(match), 6)
+	}
+
+	kubeconfig = rewriteKubeconfig(kubeconfigExample, ip, "")
+
+	re = regexp.MustCompile(`default`)
+	match = re.FindAllIndex(kubeconfig, -1)
+
+	if len(match) != 6 {
```
Comment (alexellis) [inline pkg/cmd/install_test.go]: Same as above

File: pkg/cmd/install_test.go [hunk 4]
```diff
@@ -40,8 +63,38 @@ func Test_loadPublickeyEncrypted(t *testing.T) {
...
+		t.Errorf("Unexpected error, got: %q, want: %q.", len(match), 6)
```
Comment (alexellis) [inline pkg/cmd/install_test.go]: Same magic number "6"

**Comment by alexellis** [comment]
@blackjid PTAL and also please rebase the commit (make sure you do not do a merge commit, because they cannot be accepted)

**Comment by blackjid** [comment]
What do you think now?

**Comment by alexellis** [comment]
I think that's much better. My only concern is that whilst regex may be more performant in the critical path of a webserver, strings.Contains is far more readable and maintainable in a unit test. 

I also think the names could be more concise, in OpenFaaS we prefer "want"/"got" (as more idiomatic Golang) to expected/actual. 

Thank you for revisiting the PR.

**Pull request merged**